### PR TITLE
add(metrics deliverables parser)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

Adds support to read in the metrics deliverables YAML file. This is one of likely many PRs to integrate Arnold and CG. 

This PR introduces draft functionality to setup and test the flow to arnold and will be finalised and restructured continously.

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+def generate_metrics_deliverables_model(
+    sample_id: str, sample_metrics: dict
+) -> MIPDNAMetricsDeliverables:
+    """Generate the metrics deliverable model for a sample."""
+    sample_metrics["id"] = sample_id
+    return MIPDNAMetricsDeliverables.model_validate(sample_metrics)
+
+
+def get_metrics_deliverables_file_path(pipeline: str, case_id: str) -> Path:
```
Comment (ChrOertlin) [inline cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:39]: I need to think of a way to improve the pipelines dict suggestions welcome

File: tests/apps/deliverables_metrics_parser/parser/test_parser.py
Comment (ChrOertlin) [inline tests/apps/deliverables_metrics_parser/parser/test_parser.py:1]: will add some more tests

File: cg/apps/deliverables_metrics_parser/models/pipeline_metrics_deliverables.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+"""Module for the pipeline metrics deliverables models."""
+from pydantic import BaseModel, Field
+
+
+class MIPDNAMetricsDeliverables(BaseModel):
```
Comment (ivadym) [inline cg/apps/deliverables_metrics_parser/models/pipeline_metrics_deliverables.py:5]: It'd be good to merge this model with `MetricsDeliverables`, including layers for common and pipeline-specific metrics, along with a Pydantic model for QC checks. Unsure if both implementations are necessary. 

Then we'll just have a simple static model for all metrics and use submodels to characterise them. But, as you mentioned in the code, I suppose we'll need to decide on deliverable formatting first (?).

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py (hunk 1 above)
Comment (diitaz93) [reply cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:39]: the `IntermediateDirectories` could be in constants

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        metrics_models.append(
+            generate_metrics_deliverables_model(sample_id=sample_id, sample_metrics=sample_metrics)
+        )
+    return metrics_models
+
+
+def parse_metrics_deliverables_file(pipeline: str, case_id: str) -> list[MIPDNAMetricsDeliverables]:
+    """Parse the metrics deliverables file."""
+    file_path: Path = get_metrics_deliverables_file_path(pipeline=pipeline, case_id=case_id)
+    content: list[dict[list[dict]]] = read_metrics_deliverables(file_path)
```
Comment (diitaz93) [inline cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:119]: complete the type hint for the dictionaries

File: tests/apps/deliverables_metrics_parser/parser/test_parser.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        "cg.apps.deliverables_metrics_parser.parser.deliverables_parser.get_metrics_deliverables_file_path",
+        return_value=mip_dna_metrics_deliverables_file_path,
+    )
+
+    # THEN a list of metrics deliverables models is returned
+    parsed_metrics: list[MIPDNAMetricsDeliverables] = parse_metrics_deliverables_file(
+        pipeline=Pipeline.MIPDNA, case_id="some_case"
+    )
+
+    assert isinstance(parsed_metrics[0], MIPDNAMetricsDeliverables)
```
Comment (diitaz93) [inline tests/apps/deliverables_metrics_parser/parser/test_parser.py:71]:
```suggestion
    # GIVEN a file path to a metrics deliverables file
    mocker.patch(
        "cg.apps.deliverables_metrics_parser.parser.deliverables_parser.get_metrics_deliverables_file_path",
        return_value=mip_dna_metrics_deliverables_file_path,
    )

    # WHEN parsing the metrics deliverables file
    parsed_metrics: list[MIPDNAMetricsDeliverables] = parse_metrics_deliverables_file(
        pipeline=Pipeline.MIPDNA, case_id="some_case"
    )

    # THEN a list of metrics deliverables models is returned
    assert isinstance(parsed_metrics[0], MIPDNAMetricsDeliverables)
```

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py (hunk 1 above)
Comment (islean) [reply cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:39]: I like this, although the `path_root` and `file_suffix` should be (and the root might already be) in constants.

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    else:
+        raise ValueError(f"Invalid pipeline name: {pipeline}")
+
+
+def read_metrics_deliverables(file_path: Path) -> list[dict]:
+    """Read the metrics deliverables file."""
+    return ReadFile.get_content_from_file(file_format=FileFormat.YAML, file_path=file_path)
+
+
+def get_metrics_from_content(content: list[dict[list[dict]]]) -> list[dict]:
```
Comment (islean) [inline cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:60]: Could we create a model or data class for the list[dict[list[dict]]] object?

**Review by ivadym — COMMENTED** [review]
I've only looked at the models. I think it would be a good idea to define the structure we'll be using to parse the metrics as early as possible (and perhaps perform QC validation as well)

File: cg/constants/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+"""Pipelines enum"""
+
+
+class Pipeline:
+    BALSAMIC: str = "balsamic"
+    MIPDNA: str = "mip-dna"
+    MIPRNA: str = "mip-rna"
+    MUTANT: str = "mutant"
+    RNAFUSION: str = "rna-fusion"
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/constants/pipeline.py:9]: I did not see it there! Will use that one instead!

File: cg/constants/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+"""Pipelines enum"""
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/constants/pipeline.py:1]: I did not see it there! Will use that one instead!

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py (hunk 1 above)
Comment (ChrOertlin) [reply cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:39]: It woukd not work like this, we are missing the "proper" pipeline level for some pipelines. The folder for "mip-dna" and "mip-rna" is set to "rare-disease" for example... so there is on direct way of doing this. @karlnyr mentioned a pipeline root dir in the config. will have a look and try with that.

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+from cg.apps.deliverables_metrics_parser.models.pipeline_metrics_deliverables import (
+    MIPDNAMetricsDeliverables,
+)
+from cg.constants.constants import FileFormat
+from cg.constants.pipeline import Pipeline
+from cg.io.controller import ReadFile
+
+
+class MetricsParserConstants(Enum):
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:15]: yes, I will move these modules to constants

File: cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py (hunk 5 above)
Comment (ChrOertlin) [reply cg/apps/deliverables_metrics_parser/parser/deliverables_parser.py:60]: These are the models that these files will be parsed into. At this stage files from multiple different pipelines can be used here. 

I can think of a general way in which a model can be used here.

**Pull request closed without merging**